File: CP/LeetCode/C/169.c
// 169. Majority Element

#include <stdio.h>

int majorityElement(const int *nums, int numsSize)
{
    int count = 0;
    int possible_majority = 0;

    for (int i = 0; i < numsSize; i++)
    {
        if (count == 0)
        {
            possible_majority = nums[i];
        }

        if (nums[i] == possible_majority)
        {
            count++;
        }
        else
        {
            count--;
        }
    }

    return possible_majority;
}

int main()
{
    int nums[] = {3, 2, 3};
    int numsSize = sizeof(nums) / sizeof(nums[0]);

    int result = majorityElement(nums, numsSize);
    printf("Majority Element: %d\n", result);

    return 0;
}

File: CP/LeetCode/Python/169.py
# 169. Majority Element


class Solution:
    def majorityElement(self, nums: list[int]) -> int:
        # nums.sort()
        # return nums[(len(nums) - 1)//2]

        # Boyer-Moore Voting Algorithm
        count = 0
        candidate = 0

        for num in nums:
            if count == 0:
                candidate = num

            if num == candidate:
                count += 1
            else:
                count -= 1

        return candidate


if __name__ == "__main__":
    solution = Solution()
    print(solution.majorityElement([3, 2, 3]))  # Output: 3
